### PR TITLE
Introduce docker for test

### PR DESCRIPTION
**Pull request by Koisell**
We introduce docker which allow to test locally without special
privileges. The only missing piece is to find how to publish the docker
image.

### CODE REVIEW

File: core/src/test/resources/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+RUN useradd -ms /bin/bash fs2-ssh
+RUN echo 'fs2-ssh:password' | chpasswd
+RUN mkdir -p /home/fs2-ssh/.ssh
+RUN touch /home/fs2-ssh/.ssh/authorized_keys
+ADD nopassword.pub password.pub /home/fs2-ssh/
+RUN cat /home/fs2-ssh/nopassword.pub /home/fs2-ssh/password.pub >> /home/fs2-ssh/.ssh/authorized_keys
+
+EXPOSE 22
+CMD ["/usr/sbin/sshd", "-D"]
```
Comment (nicflores) [inline core/src/test/resources/Dockerfile]: Is there an advantage to using our own Dockerfile vs using this ssh docker hub image:
https://github.com/panubo/docker-sshd ?

File: core/src/test/scala/fs2/io/ssh/ClientSpec.scala [hunk 2]
```diff
@@ -40,15 +40,16 @@ import java.nio.file.Paths
 // to run them locally, make sure you have things from
 // the "fs2-ssh test server" entry in 1Password
 // they will only run on Travis if you push your branch to upstream
-class ClientSpec extends Specification {
+class ClientSpec extends Specification with SshDockerService {
   implicit val cs = IO.contextShift(ExecutionContext.global)
 
   val Timeout = 30.seconds
 
-  val TestHost = "ec2-54-205-215-42.compute-1.amazonaws.com"
```
Comment (nicflores) [inline core/src/test/scala/fs2/io/ssh/ClientSpec.scala:48]: this is great! we can remove turn off our ec2 instance that we have running just for these test, thanks!

**Comment by nicflores** [comment]
So, I'm totally good with this PR, my only nitpick is whether or not to use: 

https://github.com/panubo/docker-sshd

instead of managing our own Dockerfile. Any thoughts @Koisell ?

**Comment by Koisell** [comment]
I must admit I'm always on the defensive when using alpine images... Considering we are on a open source project and that Travis CI runner is itself containerised, I think we could use it.

**Comment by Koisell** [comment]
I tried it. I'm struggling on how to make the test wait for the container to be ready... I will have a closer look later

**Comment by Koisell** [comment]
Another issue with https://github.com/panubo/docker-sshd is that users seem  NOT to have password

**Comment by djspiewak** [comment]
Is there a way we can simply start the docker container, then ssh in and add a user with a password and the two key files?

File: core/src/test/resources/Dockerfile (hunk 1 above)
Comment (Koisell) [reply core/src/test/resources/Dockerfile]: The advantage will be to have more control on how we parametrized our container. To tackle specific sshd version problem or special configurations.

Comment (djspiewak) [reply core/src/test/resources/Dockerfile]: Honestly that's a pretty strong advantage. I really hope we never run into that kind of thing, but my trust of upstream Apache dependencies pretty much only goes so far as I can throw their logging framework. Nothing against Mina specifically, I just… worry…

Comment (nicflores) [reply core/src/test/resources/Dockerfile]: Ok, yeah, I'm convinced. So the only thing we should now do is enable the docker service in `.travis.yml`:

 ```
services:
  - docker
```

Then build the image and run it during the `before_install` or `install` sections. Something like this should do the trick:

```
- cd core/src/test/resources/docker
- docker build -t fs2-ssh/sshd:test .
- docker run -d -p 222:22 --name test-sshd fs2-ssh/sshd:test
```

Comment (nicflores) [reply core/src/test/resources/Dockerfile]: Forgot to mention, I think we should put all the docker items in a docker directory: `core/src/test/resources/docker`

Comment (Koisell) [reply core/src/test/resources/Dockerfile]: No need of the before install, the code can do it. Both options are available which one do you prefer ?

Comment (Koisell) [reply core/src/test/resources/Dockerfile]: I prefer the code one so it work out of the box for everyone

Comment (djspiewak) [reply core/src/test/resources/Dockerfile]: I agree that code should work out of the box for everyone, at least as much as possible. Docker kind of corrupts this a bit, but as close as we can get would be ideal.

Comment (nicflores) [reply core/src/test/resources/Dockerfile]: @Koisell, ok, I see where the code builds and runs the container, thanks. I will go ahead and merge this PR. Thanks for help Koisell.

**Review by nicflores — APPROVED** [review]

**Pull request merged**